### PR TITLE
8260467: Move well-known classes from systemDictionary.hpp to vmClasses.hpp

### PR DESCRIPTION
**Pull request by iklam**
The "well-known classes" API in [systemDictionary.hpp](https://github.com/openjdk/jdk/blob/19b6f61bd2f49f06ef6b8e9b6ecd2fd910070035/src/hotspot/share/classfile/systemDictionary.hpp#L103) is for accessing a fixed set of classes defined by the bootstrap loader. The rest of systemDictionary.hpp is for finding/defining arbitrary classes in arbitrary class loaders. For modularity, we should separate these two sets of APIs, and make them independent of each other.

The proposed API follows the existing pattern of accessing known symbols/intrinsics in [vmSymbols.hpp](https://github.com/openjdk/jdk/blob/19b6f61bd2f49f06ef6b8e9b6ecd2fd910070035/src/hotspot/share/classfile/vmSymbols.hpp#L53) and [vmIntrinsics.hpp](https://github.com/openjdk/jdk/blob/19b6f61bd2f49f06ef6b8e9b6ecd2fd910070035/src/hotspot/share/classfile/vmIntrinsics.hpp#L108):

```
#include "classfile/vmClasses.hpp"
#include "classfile/vmIntrinsics.hpp"
#include "classfile/vmSymbols.hpp" // NEW

Symbol* s = vmSymbols::java_lang_Object();
vmIntrinsics::ID id = vmIntrinsics::_linkToSpecial;
Klass* k = vmClasses::Object_klass(); // NEW
```

Note: to help the review process and limit the amount of boiler-place changes, within this RFE, we do not change existing calls like `SystemDictionary::Object_klass()` to `vmClasses::Object_klass()`. That will be done in a subtask ([JDK-8260471](https://bugs.openjdk.java.net/browse/JDK-8260471) -- [preliminary webrev](http://cr.openjdk.java.net/~iklam/jdk17/8260471-decouple-vmClasses-and-SystemDictionary/)).

Within this RFE, we temporarily subclass `SystemDictionary` from `vmClasses`, so `SystemDictionary::Object_klass()` will continue to work.

Tested with mach5: tier1, builds-tier2, builds-tier3, builds-tier4 and builds-tier5. Also locally: aarch64, arm, ppc64, s390, x86, and zero.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260467](https://bugs.openjdk.java.net/browse/JDK-8260467): Move well-known classes from systemDictionary.hpp to vmClasses.hpp

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 3d380398127b37a4f642da63aaae4abfacfbcd4f
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 3d380398127b37a4f642da63aaae4abfacfbcd4f

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2246/head:pull/2246`
`$ git checkout pull/2246`

### CODE REVIEW

**Comment by iklam** [comment]
The name `vmClasses` is in `lowerCamelCase`, which is  a deviation from the rest of the HotSpot code, where most classes have `UpperCamelCase` names. However, `vmClasses` and `vmSymbols` are often used together, like the following, so a consistent look would be nice:

```
JavaCalls::call_static(&result,
   vmClasses::MethodHandleNatives_klass(),
   vmSymbols::linkMethodHandleConstant_name(),
   vmSymbols::linkMethodHandleConstant_signature(),
   &args, CHECK_(empty));
```

File: src/hotspot/share/classfile/vmClasses.cpp [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+      InstanceKlass* k = _klasses[as_int(id)];
+      assert(k->is_shared(), "must not be replaced by JVMTI class file load hook");
+    }
+  }
+#endif
+}
+
+#if INCLUDE_CDS
+
+void vmClasses::resolve_shared_class(InstanceKlass* klass, ClassLoaderData* loader_data, Handle domain, TRAPS) {
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/vmClasses.cpp:219]: It isn't at all obvious to me why this functionality was relocated. I don't see anything vmClasses related in here. ??

**Review by dholmes-ora — APPROVED** [review]
Hi Ioi,

This first step at refactoring and renaming looks good.

Only one comment below.

Thanks,
David

File: src/hotspot/share/classfile/vmClasses.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/classfile/vmClasses.cpp:219]: `vmClasses::resolve_shared_class()` is a private method and is called from here:

```
bool vmClasses::resolve(VMClassID id, TRAPS) {
  InstanceKlass** klassp = &_klasses[as_int(id)];
...
    InstanceKlass* k = *klassp;
    resolve_shared_class(k, ....);  <---- HERE
}
```

Maybe the function name is not clear enough. It's meant to "resolve a vmClass from the CDS archive". It used to be the (even more confusingly named) `SystemDictionary::quick_resolve()` method.

**Review by coleenp — COMMENTED** [review]
Why not consolidate vmClassMacros.hpp into vmClassID.hpp ?

**Comment by iklam** [comment]
> Why not consolidate vmClassMacros.hpp into vmClassID.hpp ?

ciEnv.hpp uses the VM_CLASSES_DO macro, but does not use anything inside vmClassID.hpp. As a result, vmClassMacros.hpp is included by 468 .o files, but vmClassID.hpp is included only by 318 files.

**Review by coleenp — APPROVED** [review]
Ok!

**Comment by iklam** [comment]
Thanks @coleenp and @dholmes-ora for the review!
/integrate

**Pull request closed without merging**